subprojects/execution/src/main/java/org/gradle/internal/execution/history/impl/FingerprintMapSerializer.java

```diff
@@ -50,13 +50,13 @@ public FingerprintMapSerializer(StringInterner stringInterner) {
     @Override
     public Map<String, FileSystemLocationFingerprint> read(Decoder decoder) throws IOException {
         int fingerprintCount = decoder.readSmallInt();
-        Map<String, FileSystemLocationFingerprint> fingerprints = new LinkedHashMap<String, FileSystemLocationFingerprint>(fingerprintCount);
+        ImmutableMap.Builder<String, FileSystemLocationFingerprint> fingerprints = ImmutableMap.builder();
```

Comment:
How about `builderWithExpectedSize(fingerprintCount)`?

Reply:
Does this reduce heap usage?
Does it rely on specialized `immutable` maps for `0`, `1`, `2` elements?
If that is the case I would suggest adding the relevant comment above.

Reply:
I don't think this deserves a special comment. This data is supposed immutable, so it makes no sense to use a mutable map. Plus, LinkedHashMap is a well-known memory hog due to its linked structure compared to the efficient array structure of an immutable map.